lib/adapters/autocomplete-adapter.ts

```diff
@@ -22,6 +24,18 @@ import {
 import * as ac from 'atom/autocomplete-plus';
 import { Suggestion, TextSuggestion, SnippetSuggestion } from 'atom-ide';
 
+/**
+ * Defines the behavior of suggestion acceptance.
+ * Assume you have "cons|ole" in the editor (`|` is the cursor position)
+ * and the autocomplete suggestion is `const`.
+ * - insert mode: const|ole
+ * - replace mode: const|
+ */
+enum InsertMode {
+  insert = 'insert',
+  replace = 'replace',
+}
```

Comment:
An enum with only two modes is a sign that we should just use a boolean instead.

Reply:
Not really. Neither of these options is more important, and `not insert` does not imply `replace` even in terms of meaning, so I'd argue that for code clarity it'd be best to keep the explicit enum

Reply:
> `not insert` does not imply `replace`

It actually does according to the code:

```js
if (insertMode === InsertMode.insert) {
  range = textEdit.insert;
} else {
  range = textEdit.replace;
}
```

Reply:
Yeah, I meant in terms of meaning. Having it a boolean variable doesn't help to understand what "not insert" means without looking at the code

Reply:
Using `enum` actually makes the code more complex. If you want clarity, you can add more documentation for the parameter. You can name the boolean value as `shouldReplace`. If it is false, it means to insert the text.

Reply:
> Using `enum` actually makes the code more complex.

How so, apart from more to type?

Reply:
Instead of passing a simple `1-bit Boolean`, you now require using the Enum which is an object of strings stored on the heap.

Reply:
Well, computational complexity, yes, but we're using JavaScript, a high-level language. It's not that we should write inefficient code, it's just that these kinds of inefficiencies are a drop to the ocean. And it's up to the JS runtime to perform optimizations for this kind of code. So I think that for high-level languages, code clarity is the priority.

But for code clarity, how are booleans better than enums?

P.S. I believe that booleans are not 1-bit, but at least 1 byte to be able to be efficiently addressed and not break alignment.

Reply:
Please, don't make changes that are being discussed before an agreement is reached. It's not very nice as it implies that the decision was unilaterally made by you, regardless of other opinions.

Reply:
I don't understand how enum has more clarity. This is a simple option with only two states. If you see the parameter name, you easily understand that you should pass `true` or `false`.

Reply:
For starters, JS doesn't support named parameters, so imagine a function call like `getSuggestions(text, true)`. When you see code like that, what do you think this `true` means? Now compare that to `getSuggestions(text, InsertMode.replace)`

Reply:
`enum` is not a JavaScript type. It is a TypeScript extension.

Reply:
True, but what of it?

Reply:
This makes `enum` inherently complex. JavaScript developers don't have any notion of `enum`, and so this is just making things complex for an option that could be a simple `boolean`.

Even if there were multiple states, the Union type of TypeScript was the simpler and clearer type to work with.

Reply:
We're writing code in TypeScript, not leveraging its functionality for the sake of "pure-JS devs" doesn't sound reasonable. And what about the named parameters issue?

Reply:
The named parameter is another story (more general than boolean vs enum). 

The solution is to insert documentation whenever you want to make the callsite more verbose.
```js
getSuggestions(/* text */ text, /* insertMode */ true)
```

I have been thinking to make an Atom package for this purpose.

Reply:
Yeah, but frankly this looks awful to me, and is a huge burden in terms of navigating code. It is done well in JetBrains IDEs that kind of "inject" the name into the code without it actually being there, but that's IDE-specific, and we don't want to write code that is only easy to manage in one particular kind of IDE. 

That's why the call clarity is a very strong argument in favor of the `enum`, in my opinion. Maybe we could get a third-person view at this?

Reply:
>  a very strong argument in favor of the enum,

That is not "very strong" to me. I gave you stronger arguments why this is not preferred. 

Regarding the named parameters, we have been already using those:
https://github.com/atom-community/atom-ide-outline/blob/77c5e26519d18d026320cd66bd97b98f8c25c1b8/src/main.ts#L39

> Maybe we could get a third-person view at this?

We have more important things to work on. If we want to hold an election for every single review, nothing will go forward.

Reply:
Well, your arguments are not strong to me. Seeing as you already changed the code to booleans, fine, you can keep it that way, but I must confess that this was yet another drop of demotivation for me to continue working on this. I don't think I'll be doing the tests for this PR. Feel free to add them.

Reply:
What I have learned by contributing to hundreds of projects on GitHub is to follow the coding paradigms/conventions that a specific project is using. This has given my PRs a very high chance of getting approved. No matter if I am an OOP guy or a functional guy, or if I like TypeScript enums more than booleans, I know how to follow what the reviewers want. At the end of the day, it is the "main contribution" that matters the most.

Tony Van Eerd nails this "People already have enough paradigms" greatly.
https://youtu.be/GPP64opjy_Y?t=1668